File: scm-csharp/scheme/primitives/PrimitiveCharEqP.cs
namespace scheme;

public class PrimitiveCharEqP : Primitive
{
    public override string Name()
    {
        return "char=?";
    }

    public override string Info()
    {
        return
            "Syntax: (char=? char1 char2 char3 ...)\n" +
            "Library: (scheme base)\n" +
            "Description: Returns #t if all the given characters are the same (case-sensitive comparison).\n" +
            "Example:\n" +
            "  (char=? #\\a #\\a) => #t\n" +
            "  (char=? #\\a #\\A) => #f";
    }
    
    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 2, -1);
        object current = arguments[0];
        for (int i = 1; i < arguments.Length; i++)
        {
            object next = arguments[i];
            if (!Value.AsChar(current).Equals(Value.AsChar(next))) return false;
            current = next;
        }
        return true;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveCharEqP.java
package scheme.primitives;

import scheme.Primitive;
import scheme.Value;
import scheme.SourcePos;

public class PrimitiveCharEqP extends Primitive {
    @Override
    public String name() {
        return "char=?";
    }

    @Override
    public String info() {
        return "Syntax: (char=? char1 char2 char3 ...)\n" +
               "Library: (scheme base)\n" +
               "Description: Returns #t if all the given characters are the same (case-sensitive comparison).\n" +
               "Example:\n" +
               "  (char=? #\\a #\\a) => #t\n" +
               "  (char=? #\\a #\\A) => #f";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 2, -1);
        Object current = arguments[0];
        for (int i = 1; i < arguments.length; i++) {
            Object next = arguments[i];
            if (Value.asChar(current).charValue() != Value.asChar(next).charValue()) return false;
            current = next;
        }
        return true;
    }
}
